### PR TITLE
[ONNX] Implement Attention-23

### PR DESCRIPTION
**Pull request by justinchuby**
Implement Attention-23 using sdpa.

- I used copilot for this.
- Also updated the conversion logic to remove trailing None inputs.

@gramalingam @kunal-vaishnavi @titaiwangms

### CODE REVIEW

File: torch/onnx/ops/__init__.py [hunk 1]
```diff
@@ -190,7 +191,7 @@ def symbolic_multi_out(
     *,
     dtypes: Sequence[torch.dtype | int],
     shapes: Sequence[Sequence[int | torch.SymInt]],
-    version: int | None = None,
+    version: int | None,
```
Comment (justinchuby) [inline torch/onnx/ops/__init__.py]:
```suggestion
    version: int | None = None,
```

File: torch/onnx/ops/_impl.py [hunk 2]
```diff
@@ -110,3 +111,313 @@ def rotary_embedding(
...
+    K: torch.Tensor,
+    current_q_num_heads: int,
+    current_kv_num_heads: int,
+    scale: Optional[float],
+) -> torch.Tensor:
+    """Helper function to compute QK output for qk_matmul_output_mode == 0."""
+    # Handle GQA manually for QK output
+    K_for_qk = K
+    enable_gqa = current_q_num_heads != current_kv_num_heads
+    if enable_gqa:
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply torch/onnx/ops/_impl.py]: Done

File: torch/onnx/ops/_impl.py [hunk 3]
```diff
@@ -110,3 +111,313 @@ def rotary_embedding(
...
+            # Map ONNX data type to torch dtype
+            onnx_to_torch_dtype = {
+                1: torch.float32,  # FLOAT
+                10: torch.float16,  # FLOAT16
+                11: torch.double,  # DOUBLE
+                16: torch.bfloat16,  # BFLOAT16
+            }
+            if softmax_precision in onnx_to_torch_dtype:
+                original_dtype = qk_with_bias.dtype
+                qk_with_bias = qk_with_bias.to(onnx_to_torch_dtype[softmax_precision])
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply torch/onnx/ops/_impl.py]: Done

File: torch/onnx/ops/_dtype_mappings.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
+import torch
+
+
+ONNX_DTYPE_TO_TORCH_DTYPE: dict[int, torch.dtype] = {
```
Comment (titaiwangms) [inline torch/onnx/ops/_dtype_mappings.py:4]: Should we wrap it into a function?

File: torch/onnx/ops/_impl.py [hunk 5]
```diff
@@ -110,3 +112,318 @@ def rotary_embedding(
...
+        Q = _reshape_3d_to_4d(Q, batch_size, q_num_heads)
+        K = _reshape_3d_to_4d(K, batch_size, kv_num_heads)
+        V = _reshape_3d_to_4d(V, batch_size, kv_num_heads)
+
+    assert len(Q.shape) == 4 and len(K.shape) == 4 and len(V.shape) == 4
+
+    # Calculate scale factor if not provided
+    if scale is None:
+        q_head_size = Q.shape[3]
+        scale = 1.0 / (q_head_size**0.5)
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply torch/onnx/ops/_impl.py]: Done

**Review by titaiwangms — COMMENTED** [review]
Need a test

File: torch/onnx/ops/_dtype_mappings.py (hunk 4 above)
Comment (justinchuby) [reply torch/onnx/ops/_dtype_mappings.py:4]: Is there a benefit?

**Comment by titaiwangms** [comment]
> Need a test

Not sure how to trigger Copilot, but I guess we can ask it to follow the rotary embedding op test from https://github.com/justinchuby/pytorch/blob/main/test/onnx/ops/test_ops.py

**Comment by justinchuby** [comment]
> > Need a test
> 
> Not sure how to trigger Copilot, but I guess we can ask it to follow the rotary embedding op test from [justinchuby/pytorch@`main`/test/onnx/ops/test_ops.py](https://github.com/justinchuby/pytorch/blob/main/test/onnx/ops/test_ops.py)

:) I am now doing it on vscode

**Comment by justinchuby** [comment]
TODO: Assert the schema with torch custom op functions; assert numerical results

File: torch/onnx/ops/_dtype_mappings.py (hunk 4 above)
Comment (titaiwangms) [reply torch/onnx/ops/_dtype_mappings.py:4]: Just don't need it to be a global dict being initilaized and passed around. We can functionalize it, and only initilaize it when we need it. Like the way we use from_type() and to_type().

File: test/onnx/ops/test_ops.py [hunk 6]
```diff
@@ -514,6 +514,305 @@ def forward(
...
+
+        # Output should be reshaped back to 3D
+        self.assertEqual(output.shape, (batch_size, q_seq_len, q_num_heads * head_size))
+        self.assertEqual(present_key.shape, (batch_size, kv_num_heads, kv_seq_len, head_size))
+        self.assertEqual(present_value.shape, (batch_size, kv_num_heads, kv_seq_len, head_size))
+
+    def test_attention_gqa(self):
+        """Test Group Query Attention (GQA)."""
+        batch_size, q_seq_len, kv_seq_len = 2, 4, 6
+        q_num_heads, kv_num_heads = 8, 4  # GQA: q_num_heads > kv_num_heads
```
Comment (kunal-vaishnavi) [inline test/onnx/ops/test_ops.py]:
```suggestion
        q_num_heads, kv_num_heads = 8, 4  # GQA: q_num_heads % kv_num_heads = 0
```

File: test/onnx/ops/test_ops.py [hunk 7]
```diff
@@ -514,6 +514,305 @@ def forward(
...
+        K = torch.rand(batch_size, kv_num_heads, kv_seq_len, head_size)
+        V = torch.rand(batch_size, kv_num_heads, kv_seq_len, head_size)
+
+        output, present_key, present_value, qk_output = torch.onnx.ops.attention(
+            Q, K, V
+        )
+
+        self.assertEqual(output.shape, (batch_size, q_num_heads, q_seq_len, head_size))
+
+    def test_attention_with_mask(self):
```
Comment (kunal-vaishnavi) [inline test/onnx/ops/test_ops.py]: We should specify the attention mask's shape being tested in the function's name. Attention masks can be a wide variety of shapes.

File: torch/onnx/ops/_impl.py [hunk 8]
```diff
@@ -110,3 +113,322 @@ def rotary_embedding(
...
+    else:
+        # For other modes, return a zero tensor with correct shape
+        return torch.zeros_like(torch.matmul(Q, K.transpose(-2, -1)))
+
+
+def _validate_gqa_configuration(
+    current_q_num_heads: int, current_kv_num_heads: int
+) -> None:
+    """Validate Group Query Attention configuration."""
+    if current_q_num_heads != current_kv_num_heads:
```
Comment (kunal-vaishnavi) [inline torch/onnx/ops/_impl.py]: The if condition is not needed. The assert should be sufficient.

File: torch/onnx/ops/_impl.py [hunk 9]
```diff
@@ -110,3 +113,322 @@ def rotary_embedding(
...
+    # Update K and V to include past states
+    K, V = present_key, present_value
+
+    # Get current dimensions
+    current_q_num_heads = Q.shape[1]
+    current_kv_num_heads = K.shape[1]
+    q_sequence_length = Q.shape[2]
+    kv_sequence_length = K.shape[2]
+
+    # Check if we can use the optimized scaled_dot_product_attention (most optimized)
```
Comment (gramalingam) [inline torch/onnx/ops/_impl.py:263]: @justinchuby : is an optimized implementation here important? I see 3 implementations here. Is it worth the added complexity? Trying to understand the intended usage scenario. Would just the last one (more general implementation) suffice?

Comment (justinchuby) [reply torch/onnx/ops/_impl.py:263]: I was hoping that we can have a reasonable performance so that if the op is decomposed it takes the more efficient path. But maybe that's not very important?

Comment (justinchuby) [reply torch/onnx/ops/_impl.py:263]: I will remove the flex attention implementation

File: torch/onnx/ops/_impl.py (hunk 8 above)
Comment (justinchuby) [reply torch/onnx/ops/_impl.py]: Done

File: torch/onnx/ops/_impl.py (hunk 9 above)
Comment (justinchuby) [reply torch/onnx/ops/_impl.py:263]: Removed

File: torch/onnx/ops/_impl.py [hunk 10]
```diff
@@ -110,3 +113,257 @@ def rotary_embedding(
...
+    if len(Q.shape) == 3:
+        assert q_num_heads != 0 and kv_num_heads != 0, (
+            "q_num_heads and kv_num_heads must be provided for 3D inputs"
+        )
+        q_sequence_length = Q.shape[1]
+        Q = _reshape_3d_to_4d(Q, batch_size, q_num_heads)
+        K = _reshape_3d_to_4d(K, batch_size, kv_num_heads)
+        V = _reshape_3d_to_4d(V, batch_size, kv_num_heads)
+
+    assert len(Q.shape) == 4 and len(K.shape) == 4 and len(V.shape) == 4
```
Comment (titaiwangms) [inline torch/onnx/ops/_impl.py]: Why do we need reshape it to 4d?

Comment (titaiwangms) [reply torch/onnx/ops/_impl.py]: Is it sdpa only accepting 4d?

Comment (justinchuby) [reply torch/onnx/ops/_impl.py]: I think because torch spda requires 4d. And it is easier to work with

File: torch/onnx/ops/_dtype_mappings.py (hunk 4 above)
Comment (justinchuby) [reply torch/onnx/ops/_dtype_mappings.py:4]: I looked at this again: I think it's ok because it is internal use. This keeps the code simpler

File: test/onnx/ops/test_ops.py (hunk 7 above)
Comment (justinchuby) [reply test/onnx/ops/test_ops.py]: Please take a look again, thanks!

File: torch/onnx/ops/_impl.py [hunk 11]
```diff
@@ -110,3 +112,279 @@ def rotary_embedding(
...
+                1,  # FLOAT
+                10,  # FLOAT16
+                11,  # DOUBLE
+                16,  # BFLOAT16
+            }
+            if softmax_precision in allowed_intermediate_precisions:
+                original_dtype = qk_with_bias.dtype
+                qk_with_bias = qk_with_bias.to(
+                    _dtype_mappings.ONNX_DTYPE_TO_TORCH_DTYPE[softmax_precision]
+                )
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply torch/onnx/ops/_impl.py:375]: Done

File: torch/onnx/_internal/exporter/_core.py [hunk 12]
```diff
@@ -552,6 +552,11 @@ def _handle_call_function_node_with_lowering(
     if _is_onnx_op(node.target):
         # Handle torch.ops.onnx.* ops. These ops can be directly added to the graph
         op_type, opset_version = _parse_onnx_op(node.target)  # type: ignore[arg-type]
+        # If final inputs are None, strip them from the node inputs
```
Comment (titaiwangms) [inline torch/onnx/_internal/exporter/_core.py:555]: Why do we need this?

File: test/onnx/ops/test_ops.py [hunk 13]
```diff
@@ -512,7 +525,928 @@ def forward(
         )
         self.assertEqual(onnx_program.model.opset_imports[""], 23)
         self.assertEqual("RotaryEmbedding", onnx_program.model.graph.node(0).op_type)
-        print(onnx_program)
+
+    def test_attention_basic(self):
```
Comment (titaiwangms) [inline test/onnx/ops/test_ops.py:529]: The test naming is too general. Maybe add something like check output shape or op spec, etc.

File: test/onnx/ops/test_ops.py [hunk 14]
```diff
@@ -512,7 +525,928 @@ def forward(
...
+
+        self.assertEqual(output.shape, (batch_size, q_num_heads, q_seq_len, head_size))
+        self.assertEqual(present_key.shape, K.shape)
+        self.assertEqual(present_value.shape, V.shape)
+        torch.testing.assert_close(output, expected)
+
+    def test_attention_mqa(self):
+        """Test Multi-Query Attention (MQA)."""
+        batch_size, q_seq_len, kv_seq_len = 2, 4, 6
+        q_num_heads, kv_num_heads = 8, 1  # MQA: kv_num_heads = 1
```
Comment (titaiwangms) [inline test/onnx/ops/test_ops.py:606]: Looks like just kv_num_heads different from the previous test. The test seems not very significant to stay in CI, although it's a different Attention from a design point of view.

File: test/onnx/ops/test_ops.py [hunk 15]
```diff
@@ -512,7 +525,928 @@ def forward(
...
+            Q, K, V
+        )
+        expected = torch.nn.functional.scaled_dot_product_attention(
+            Q, K, V, None, enable_gqa=True
+        )
+
+        self.assertEqual(output.shape, (batch_size, q_num_heads, q_seq_len, head_size))
+        torch.testing.assert_close(output, expected)
+
+    def test_attention_with_2d_mask(self):
```
Comment (titaiwangms) [inline test/onnx/ops/test_ops.py:624]: It's better specify what test we are doing. For example, this one checks registration and output shape. But the previous ones actually check eager mode.

File: test/onnx/ops/test_ops.py [hunk 16]
```diff
@@ -512,7 +525,928 @@ def forward(
...
+
+        self.assertEqual(
+            output_2d.shape, (batch_size, q_num_heads, q_seq_len, head_size)
+        )
+        self.assertEqual(
+            output_4d.shape, (batch_size, q_num_heads, q_seq_len, head_size)
+        )
+
+    def test_attention_with_large_negative_float_mask(self):
+        """Test attention with large negative values in float mask."""
```
Comment (titaiwangms) [inline test/onnx/ops/test_ops.py:745]: Why is this one important? Is there a real case we need large negative values of fp mask?

**Review by titaiwangms — APPROVED** [review]
Minor comments - unblocking:

I think in general the test names can be more specific. On the other hand, from framework (exporter) perspective, I personally think some of the tests are redundant (too detail).

File: test/onnx/ops/test_ops.py (hunk 13 above)
Comment (kunal-vaishnavi) [reply test/onnx/ops/test_ops.py:529]: I agree. Something like `test_attention_without_past_kv_caches` or `test_attention_with_present_kv_caches` could work.

File: test/onnx/ops/test_ops.py (hunk 14 above)
Comment (justinchuby) [reply test/onnx/ops/test_ops.py:606]: Makes sense, will simplify

File: test/onnx/ops/test_ops.py (hunk 16 above)
Comment (justinchuby) [reply test/onnx/ops/test_ops.py:745]: Yeah I will remove. I blame copilot

File: test/onnx/ops/test_ops.py (hunk 14 above)
Comment (kunal-vaishnavi) [reply test/onnx/ops/test_ops.py:606]: Can `parameterized.expand` be used to parameterize on the value of `kv_num_heads`? Then only one function is needed.

File: test/onnx/ops/test_ops.py (hunk 15 above)
Comment (kunal-vaishnavi) [reply test/onnx/ops/test_ops.py:624]: Can all of the tests have the same checks for registration, output shape, etc?

File: test/onnx/ops/test_ops.py [hunk 17]
```diff
@@ -512,7 +525,928 @@ def forward(
...
+        output_float, _, _, _ = torch.onnx.ops.attention(Q, K, V, attn_mask=float_mask)
+
+        self.assertEqual(
+            output_bool.shape, (batch_size, q_num_heads, q_seq_len, head_size)
+        )
+        self.assertEqual(
+            output_float.shape, (batch_size, q_num_heads, q_seq_len, head_size)
+        )
+
+    def test_attention_with_zero_float_mask(self):
```
Comment (kunal-vaishnavi) [inline test/onnx/ops/test_ops.py:680]: Can this test be added to the 2D attention mask test instead? That one has a test for a boolean mask and a float mask already.

File: test/onnx/ops/test_ops.py [hunk 18]
```diff
@@ -512,7 +525,928 @@ def forward(
...
+        K = torch.rand(batch_size, kv_num_heads, kv_seq_len, head_size)
+        V = torch.rand(batch_size, kv_num_heads, kv_seq_len, head_size)
+
+        zero_mask = torch.zeros(q_seq_len, kv_seq_len)
+        torch.library.opcheck(_impl.attention_23, (Q, K, V), dict(attn_mask=zero_mask))
+        output, _, _, _ = torch.onnx.ops.attention(Q, K, V, attn_mask=zero_mask)
+
+        self.assertEqual(output.shape, (batch_size, q_num_heads, q_seq_len, head_size))
+
+    def test_attention_with_causal_mask_pattern(self):
```
Comment (kunal-vaishnavi) [inline test/onnx/ops/test_ops.py:696]: Same question for adding to the 2D attention mask test

File: test/onnx/ops/test_ops.py [hunk 19]
```diff
@@ -512,7 +525,928 @@ def forward(
...
+        # Create a lower triangular causal mask
+        causal_mask = torch.tril(torch.ones(q_seq_len, kv_seq_len, dtype=torch.bool))
+        torch.library.opcheck(
+            _impl.attention_23, (Q, K, V), dict(attn_mask=causal_mask)
+        )
+        output, _, _, _ = torch.onnx.ops.attention(Q, K, V, attn_mask=causal_mask)
+
+        self.assertEqual(output.shape, (batch_size, q_num_heads, q_seq_len, head_size))
+
+    def test_attention_with_gqa_and_mask(self):
```
Comment (kunal-vaishnavi) [inline test/onnx/ops/test_ops.py:715]: Can we generalize all of the tests to be for GQA since GQA (`num_heads % num_key_value_heads == 0`) is a superset of MHA (`num_heads = num_key_value_heads`) and MQA (`num_key_value_heads = 1`)?

File: test/onnx/ops/test_ops.py (hunk 16 above)
Comment (kunal-vaishnavi) [reply test/onnx/ops/test_ops.py:745]: ORT [uses](https://github.com/microsoft/onnxruntime/blob/7268117855a89b1ce18b1ec78078b159d3bd661c/onnxruntime/core/graph/contrib_ops/bert_defs.cc#L412) a large mask filter value to represent `-inf`.

**Comment by titaiwangms** [comment]
Feel free to merge first to catch the branch cut, since all the comments are about tests (not user facing.)

**Comment by justinchuby** [comment]
Makes sense, will do. Thanks for reviewing!

File: torch/onnx/_internal/exporter/_core.py (hunk 12 above)
Comment (justinchuby) [reply torch/onnx/_internal/exporter/_core.py:555]: We have the logic in the builder that does this. Basically if an op has optional positional arguments it can carry Nones. In onnx trailing inputs being None is equivalent to not having those inputs. However for optimizers it is often easier if we normalize it (by removing the trailing inputs) to checking input counts etc.

**Comment by justinchuby** [comment]
Follow up: https://github.com/pytorch/pytorch/pull/156635
